### DISCUSSION TITLE
Roles won't stick

### DISCUSSION THREAD

**Discussion by androidacy-user** [spatie/laravel-permission#2176] [Q&A] [answered]
This one is driving us crazy, and we're not sure what's causing it.

We're using Laravel 9.29 via Laravel Octane w/ Swoole and about every 10-20 minutes, all users get assigned the "Guest" role (basically our placeholder role for no permissions). Trying to reassign the role gets reverted about 10-20 minutes later.

There's no scheduled task that could be doing this. We've tried messing with cache and re-examined all of our auth logic to make sure errant users aren't getting reassigned roles.

We've even tried overriding the syncRoles() and assignRole() methods to try to catch what's doing it and nothing was caught but the roles were reverted.

Any ideas?

**Comment by fredsal** [comment]
> all users get assigned the "Guest" role (basically our placeholder role for no permissions)

It looks like you broke something in your implementation.

>Any ideas?

nobody could tell you which part you broke

**Reply by androidacy-user** [reply]
So very helpful, you are.

We log every action that has the ability to change these values. We've tried inserting stops into the framework itself to catch when it happens.

Nothing is catching anything.

We've even tried redoing the migrations and starting clean. Now the "revert" is happening faster. Once again, seemingly without warning.

**Reply by androidacy-user** [reply]
Not sure if it makes a difference but our Users table doesn't use id #'s for primary key but the api token of the user

**Reply by fredsal** [reply]
> So very helpful, you are.

if you want better help you should post the sources or a failing test on a minimal app on fresh instalation to demonstrate the problem, 
you won't be thinking someone is going to use their telepathic powers to tell you which line you broke the functionality on

**Reply by androidacy-user** [reply]
> > So very helpful, you are.
> 
> if you want better help you should post the sources or a failing test on a minimal app on fresh instalation to demonstrate the problem, you won't be thinking someone is going to use their telepathic powers to tell you which line you broke the functionality on

"you broke something" is that all you can say?

I did figure it out. It's an undocumented requirement, apparently. See my answer below.

**Comment by androidacy-user** [comment, marked as answer]
So after altering the table to have an auto incrementing `id` column (whereas before the `token` column, which had the unique constraint but not the auto increment one), the issue went away. Also resolved another issue where it would complain role rows weren't unique.

Unless, perhaps, I missed it in the docs somewhere looks like either a bug or undocumented intended behavior.
